torch/csrc/autograd/record_function.h

```diff
@@ -93,49 +93,49 @@ struct TORCH_API RecordFunction {
     return inputs_;
   }
 
-  inline const RecordFunction* parent() const {
-    return parent_;
-  }
-
   bool active() const {
     return initialized_;
   }
 
-  void setRunSampled(bool run_sampled) {
+  // Internal, only for the use within RECORD_FUNCTION macro;
+  // enables this record function to run sampled callbacks
+  void _setRunSampled(bool run_sampled) {
     run_sampled_ = run_sampled;
   }
 
-  void end();
+  // Internal, only for the use within RECORD_FUNCTION macro;
+  // sets this function as the current() thread local function;
+  // original value of current() is restored in destructor
+  void _setCurrent();
 
-  // Saves the thread_id that this RecordFunction was created with. This is
-  // needed so that we can access Events created by the original thread in a
-  // different thread, since they are thread-local. This should be used to call
-  // RecordFunction::end() in a different thread.
-  void setThreadId();
+  // Executes end callbacks
+  void end();
 
   // Retrieves the thread_id that this RecordFunction was created with. Useful
   // if we need to access Events created by the original thread in a different
-  // thread. The threadId_ should only be set (via setThreadId) in cases where
-  // RecordFunction::end is called in a different thread.
+  // thread.
   inline uint16_t getThreadId() const {
     return threadId_;
   }
 
+  // Get thread_id for the current thread
+  static uint16_t getCurrentThreadId();
+
  private:
   void processCallbacks();
 
   Node* fn_ = nullptr;
   StringView name_;
   int64_t sequence_nr_ = -1;
   std::vector<c10::IValue> inputs_;
-  // parent_ points to the parent RecordFunction and must out live this.
+  // parent_ points to the parent RecordFunction and must out live this;
+  // only to be used together with RECORD_FUNCTION macro
   RecordFunction* parent_ = nullptr;
 
   bool initialized_ = false;
   bool run_sampled_ = false;
-  // The thread_id that this RecordFunction was created with. If 0, this means
-  // that it was not set with setThreadId() and this RecordFunction's callbacks
-  // cannot be invoked from a separate thread.
+  bool is_current_ = false;
```

Comment:
nit: add a comment indicating that this flag is only true in the RECORD_FUNCTION macro for scope management?